### DISCUSSION TITLE
Generic types aren't properly parsed via FastAPI APIRouter

### DISCUSSION THREAD

**Discussion by walter9388** [fastapi/fastapi#6062] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Description

I have a fairly complex base class in an app which I want to reuse throughout for ease of implementation and readability. However, the types of the input arguments to each of the underlying API routes changes depending on the route. To summarise the issue, I have made a very basic implementation with similar logic below:

Filename=`test.py`:
```python
from typing import Generic, List, TypeVar
from fastapi import APIRouter, FastAPI

app = FastAPI()

T = TypeVar("T")

class Test(Generic[T]):
    def __init__(self, route_base_name: str):
        # create routes
        self.router = APIRouter()
        self.router.add_api_route(
            f"/{route_base_name}/", self.test, methods=["POST"], response_model=str
        )

    def test(self, x: T):
        print(type(x))
        return str(type(x))

app.include_router(Test[int]("test1").router)
app.include_router(Test[List[int]]("test2").router)
app.include_router(Test[List[str]]("test3").router)

if __name__ == "__main__":
    import uvicorn

    uvicorn.run("test:app", host="0.0.0.0", port=3001, reload=True)
```
To run: `python test.py`.

If we look at the swagger documentation from the above code, we see that type hinting has been ignored for all the routes, and they all expect type `str` (test3 below should expect `List[str]` type etc.): 
![image](https://user-images.githubusercontent.com/51122673/211634128-1a6954df-e409-435c-a465-ea3b21e83013.png)

What I was expecting to see is the equivalent to if I write out the above code out fully.
Filename=`test2.py`:
```python
from typing import List
from fastapi import APIRouter, FastAPI


app = FastAPI()


class Test1:
    def __init__(self, route_base_name: str):
        # create routes
        self.router = APIRouter()
        self.router.add_api_route(
            f"/{route_base_name}/", self.test, methods=["POST"], response_model=str
        )

    def test(self, x: int):
        print(type(x))
        return str(type(x))


class Test2:
    def __init__(self, route_base_name: str):
        # create routes
        self.router = APIRouter()
        self.router.add_api_route(
            f"/{route_base_name}/", self.test, methods=["POST"], response_model=str
        )

    def test(self, x: List[int]):
        print(type(x))
        return str(type(x))


class Test3:
    def __init__(self, route_base_name: str):
        # create routes
        self.router = APIRouter()
        self.router.add_api_route(
            f"/{route_base_name}/", self.test, methods=["POST"], response_model=str
        )

    def test(self, x: List[str]):
        print(type(x))
        return str(type(x))


app.include_router(Test1("test1").router)
app.include_router(Test2("test2").router)
app.include_router(Test3("test3").router)

if __name__ == "__main__":
    import uvicorn

    uvicorn.run("test2:app", host="0.0.0.0", port=3001, reload=True)

```
To run: `python test2.py`.

This time when we look at the swagger documentation, the routes have the expected type hinting:
![image](https://user-images.githubusercontent.com/51122673/211634767-0cfd8d78-b753-4d63-a92f-ffb04ea628ac.png)

I think this is a FastAPI issue (although my understanding of python generic types may be off). I would appreciate any advice on how to solve this problem without needing to write out all the code fully each time.

Thanks in advance,
Alex

### Operating System

Linux

### Operating System Details

Ubuntu 20.04.4 LTS

### FastAPI Version

0.78.0

### Python Version

3.8.10

### Additional Context

I have read through the [FastAPI typing documentation](https://fastapi.tiangolo.com/python-types/#generic-types) and couldn't find a solution which fitted. I have also seen similar stuff from the pydantic side (e.g. [here](https://github.com/tiangolo/fastapi/issues/653#issuecomment-984509798)), however this also doesn't seem to fit the simplicity of the problem I am having as I don't think I should need to use pydantic explicitly / at all in the above example.

**Comment by rra** [comment]
I just encountered what I think is the same or a similar problem but with response models.

I have a set of routes as part of a generic library that return a parameterized generic type as the response model (`Job[T]`), which contain a field typed as `parameters: T`. Pydantic works correctly all the way down. In the handler, if I call `.dict()` on the `Job[T]` object I'm about to return, it includes the `parameters` key and the value is correctly serialized based on the parameterization `T`. But if I return that object and let Pydantic serialize it to JSON, the `parameters` field is missing entirely.

This sounds very much like the symptoms of the Pydantic issue linked above, but the `Job` class does inherit from `GenericModel` and as mentioned Pydantic itself (via `.dict()` or `.json()`) has no trouble. Apparently FastAPI is doing something more complex than calling `.dict()` on the returned object, and whatever that is is going wrong in this case.

**Comment by walter9388** [comment]
Sounds like a similar issue. Hopefully @tiangolo or someone from the dev crew can provide some feedback sometime soon 😊

**Comment by rra** [comment, marked as answer]
I think I now understand this problem and I think both of the problems we encountered are versions of the same thing. I'm also not sure if Python provides enough information for FastAPI to support this the way that we would prefer.

The basic problem is that FastAPI relies on the type information in the type annotations to decide how to decode input and encode output. It doesn't use the runtime types of objects as far as I can tell, only the static types from the type declarations, with the exception of a `response_model` parameter. (This is probably a simplification, but it's consistent with my experiments.) But with generics the type annotation is a `TypeVar` with a bound, and that's not resolved to a concrete type in the type annotation. Therefore, FastAPI has only the `TypeVar` to work with and therefore treats it as equivalent to its bound, rather than the more specialized type that it actually represents.

The workaround I found for this is pretty ugly since it involves poking at the internals of the type annotations, but if you adjust the type annotation to have the correct concrete type, FastAPI then picks that up and the right things happen. But you have to write the code differently, since you need to create a separate copy of the function for each different type signature you want it to have.

The following test program based on your test program works as expected. I chose to explicitly pass in the desired type to the constructor, since that makes things fairly easy, but I suspect you could recover it from the type parameter with similar deep magic if you wanted to.

```python
from typing import Generic, List, TypeVar
from fastapi import APIRouter, FastAPI

app = FastAPI()

T = TypeVar("T")

class Test(Generic[T]):
    def __init__(self, route_base_name: str, body_type: T):
        def test(x: T):
            print(type(x))
            return str(type(x))

        test.__annotations__["x"] = body_type

        # create routes
        self.router = APIRouter()
        self.router.add_api_route(
            f"/{route_base_name}/", test, methods=["POST"], response_model=str
        )

app.include_router(Test("test1", int).router)
app.include_router(Test("test2", List[int]).router)
app.include_router(Test("test3", List[str]).router)

if __name__ == "__main__":
    import uvicorn

    uvicorn.run("test:app", host="0.0.0.0", port=3001, reload=True)
```

**Comment by walter9388** [comment]
Brilliant! Thank you @rra!

Yeah, it's a bit of a clunky solution, but this work around actually makes a lot of sense to me now I've thought about it, and it is exactly the sort of solution I was looking for tbh. It still seems to me like there might be a way for FastAPI to incorporate some similar logic in the background to stop this extra boilerplate code being needed, but this will certainly do for now. 
Thank you for also explaining the issue as well, it definitely helped me understand the problem.

Just as a minor addition, I need to be able to overwrite the internal `test` function in some of the children of the class `Test`. Clearly your solution still holds, but I thought I would just amend it to the following for completeness.

```python
from typing import Generic, List, TypeVar
from fastapi import APIRouter, FastAPI

app = FastAPI()

T = TypeVar("T")

class Test(Generic[T]):
    def __init__(self, route_base_name: str, body_type: T):
        # boilerplate for generics parsing
        def _test(x: T):
            self.test(x)
        _test.__annotations__["x"] = body_type

        # create routes
        self.router = APIRouter()
        self.router.add_api_route(
            f"/{route_base_name}/", _test, methods=["POST"], response_model=str
        )

    def test(self, x: T):
        print(type(x))
        return str(type(x))

app.include_router(Test("test1", int).router)
app.include_router(Test("test2", List[int]).router)
app.include_router(Test("test3", List[str]).router)

if __name__ == "__main__":
    import uvicorn

    uvicorn.run("test:app", host="0.0.0.0", port=3001, reload=True)
```

Thanks again 😊

**Comment by ahmed-aly-aut** [comment]
Hello,
I found this issue when I was struggeling with the same issue. My project is setup with python==3.12.1 and fastapi==0.108.0.
I found possible solution for this. Debugging the created routes I found out, that the Generics were not resolved as I would have hoped and with the information I got from this discussion I found a solution getting the actual Generic types at runtime with `get_args(self.__orig_bases__[0])` and using it in the code.

This works for me:

```python
from collections.abc import Sequence
from typing import Annotated, Literal, get_args

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, NonNegativeInt
from sqlalchemy.orm import exc
from sqlmodel import SQLModel

from app.services.base_service import BaseService


class GenericRouteGenerator[M: SQLModel, CS: BaseModel, US: BaseModel, RS: BaseModel]:

    def __init__(self, service: BaseService[M, CS, US], router: APIRouter) -> None:
        self.service = service
        self.router = router

        # Dynamically retrieve generic type arguments
        model, create_schema, update_schema, read_schema = get_args(self.__orig_bases__[0])

        # Set annotations for FastAPI to pick up
        self.__annotations__ = {
            "model": model,
            "create_schema": create_schema,
            "update_schema": update_schema,
            "read_schema": read_schema
        }

    def add_create_route(self) -> None:
        create_schema = self.__annotations__["create_schema"]
        read_schema = self.__annotations__["read_schema"]
        async def create_object(create_model: create_schema) -> M:
            return await self.service.create(create_model)

        self.router.add_api_route("/", create_object, methods=["POST"], response_model=read_schema)

    def add_read_route(self) -> None:
        read_schema = self.__annotations__["read_schema"]
        async def read_object(obj_id: int) -> M:
            obj = await self.service.read(obj_id)
            if not obj:
                raise HTTPException(status_code=404, detail="Ressource not found")
            return obj

        self.router.add_api_route("/{obj_id}", read_object, methods=["GET"], response_model=read_schema)

    def add_read_all_route(self) -> None:
        read_schema = self.__annotations__["read_schema"]
        async def read_objects(
            offset: Annotated[NonNegativeInt | None, Query()] = 0,
            limit: Annotated[NonNegativeInt | None, Query()] = 10,
            order: Annotated[str | None, Query()] = None,
        ) -> Sequence[M]:
            return await self.service.read_all(offset, limit, order)

        self.router.add_api_route("/", read_objects, methods=["GET"], response_model=list[read_schema])

    def add_patch_route(self) -> None:
        update_schema = self.__annotations__["update_schema"]
        read_schema = self.__annotations__["read_schema"]
        async def update_object(obj_id: int, update_data: update_schema) -> M:
            obj = await self.service.update(obj_id, update_data)
            if not obj:
                raise HTTPException(status_code=404, detail="Resource not found")
            return obj

        self.router.add_api_route("/{obj_id}", update_object, methods=["PATCH"], response_model=read_schema)

    def add_delete_route(self) -> None:
        async def delete_object(obj_id: int) -> dict[str, Literal[True]]:
            try:
                await self.service.delete(obj_id)
                return {"ok": True}
            except exc.UnmappedInstanceError as err:
                raise HTTPException(
                    status_code=404, detail="Country not found"
                ) from err

        self.router.add_api_route("/{obj_id}", delete_object, methods=["DELETE"])

    def get_router(self) -> APIRouter:
        return self.router

```

Type checking with tools like mypy is a bit useless right now because mypy doesn't support PEP 695 generics yet.

**Reply by ahmed-aly-aut** [reply]
And you need something like that.

```python
router = APIRouter(
    responses={404: {"description": "Not found"}},
)

address_repository = ....
address_service = AddressService(address_repository)

class AddressRouter(GenericRouteGenerator[Address, AddressCreate, AddressUpdate, AddressRead]):
    pass

address_route_generator = AddressRouter(
    address_service,
    router
)
```
The class AddressRouter will make it work correctly

**Reply by flo-at** [reply]
Since Python 3.12 there is [types.get_original_bases](https://docs.python.org/3.13/library/types.html#types.get_original_bases) so you don't have to rely on `self.__orig_bases__`.

Your suggestion gets me quite close to what I'm looking for.
My equivalent to your line `async def create_object(create_model: create_schema) -> M:` results in a `pydantic.errors.PydanticUndefinedAnnotation: name 'create_schema' is not defined` for me though.
I can print `create_schema` right before that to verify the correct type is stored there so the schema type is definitley there..
Python 3.12.2 and Pydantic 2.6.1.

> Type checking with tools like mypy is a bit useless right now because mypy doesn't support PEP 695 generics yet.

Your suggestion should also work for the "old" Generic syntax, right?

Edit:
It works for me if I add
```py
create_object.__annotations__["create_model"] = create_schema
```
like @walter9388 did in his comment.

**Comment by ahmed-aly-aut** [comment]
After some experimentation and using some design patterns, I was able to  improve on the code i shared earlier.
Here is my final solution and thanks @flo-at for the tip with [types.get_original_bases](https://docs.python.org/3.12/library/types.html#types.get_original_bases):

```python
class GenericRouteGenerator[
    M: SQLModel,
    CS: BaseModel,
    US: BaseModel,
    RS: BaseModel,
    S: AsyncBaseService,
    R: AsyncBaseRepository,
]:
    def __init__(self, router: APIRouter) -> None:
        self.router = router

        # Dynamically retrieve generic type arguments
        (
            model,
            create_schema,
            update_schema,
            read_schema,
            service,
            repository,
        ) = get_args(get_original_bases(type(self))[0])

        # Set annotations for FastAPI to pick up
        self.__annotations__ = {
            "model": model,
            "create_schema": create_schema,
            "update_schema": update_schema,
            "read_schema": read_schema,
            "service": service,
            "repository": repository,
        }

    def add_create_route(self) -> None:
        create_schema = self.__annotations__["create_schema"]
        read_schema = self.__annotations__["read_schema"]
        _repository = self.__annotations__["repository"]
        _service = self.__annotations__["service"]

        async def create_object(
            create_model: create_schema,
            service: Annotated[S, Depends(get_service(_service, _repository))],
        ) -> M:
            obj = await service.create(create_model)
            if not obj:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Error creating resource",
                )
            return obj

        self.router.add_api_route(
            "/",
            create_object,
            methods=["POST"],
            response_model=read_schema,
            status_code=status.HTTP_201_CREATED,
        )

    def add_read_route(self) -> None:
        read_schema = self.__annotations__["read_schema"]
        _repository = self.__annotations__["repository"]
        _service = self.__annotations__["service"]

        async def read_object(
            obj_id: int,
            service: Annotated[S, Depends(get_service(_service, _repository))],
        ) -> M:
            obj = await service.read(obj_id)
            if not obj:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND, detail="Ressource not found"
                )
            return obj

        self.router.add_api_route(
            "/{obj_id}", read_object, methods=["GET"], response_model=read_schema
        )

    def add_read_all_route(self) -> None:
        read_schema = self.__annotations__["read_schema"]
        _repository = self.__annotations__["repository"]
        _service = self.__annotations__["service"]

        async def read_objects(
            *,
            skip: Annotated[NonNegativeInt | None, Query()] = 0,
            limit: Annotated[NonNegativeInt | None, Query()] = 10,
            order: Annotated[str | None, Query()] = None,
            service: Annotated[S, Depends(get_service(_service, _repository))],
        ) -> Sequence[M]:
            return await service.read_all(skip, limit, order)

        self.router.add_api_route(
            "/", read_objects, methods=["GET"], response_model=list[read_schema]
        )

    def add_patch_route(self) -> None:
        update_schema = self.__annotations__["update_schema"]
        read_schema = self.__annotations__["read_schema"]
        _repository = self.__annotations__["repository"]
        _service = self.__annotations__["service"]

        async def update_object(
            obj_id: int,
            update_data: update_schema,
            service: Annotated[S, Depends(get_service(_service, _repository))],
        ) -> M:
            obj = await service.update(obj_id, update_data)
            if not obj:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND, detail="Resource not found"
                )
            return obj

        self.router.add_api_route(
            "/{obj_id}", update_object, methods=["PATCH"], response_model=read_schema
        )

    def add_delete_route(self) -> None:
        _repository = self.__annotations__["repository"]
        _service = self.__annotations__["service"]

        async def delete_object(
            obj_id: int,
            service: Annotated[S, Depends(get_service(_service, _repository))],
        ) -> None:
            try:
                await service.delete(obj_id)
            except exc.UnmappedInstanceError as err:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND, detail="Country not found"
                ) from err

        self.router.add_api_route(
            "/{obj_id}",
            delete_object,
            methods=["DELETE"],
            status_code=status.HTTP_204_NO_CONTENT,
        )

    def add_all_routes(self) -> None:
        self.add_create_route()
        self.add_read_route()
        self.add_read_all_route()
        self.add_patch_route()
        self.add_delete_route()

    def get_router(self) -> APIRouter:
        return self.router

```

And using it:

```python
class AddressRouter(
    GenericRouteGenerator[
        Address,
        AddressCreate,
        AddressUpdate,
        AddressRead,
        AddressService,
        AddressRepository,
    ]
):
    pass


address_route_generator = AddressRouter(router)

address_route_generator.add_all_routes()

app = FastAPI()
app.include_router(address_route_generator.get_router())
```

**Reply by flo-at** [reply]
That's similar to what I ended up doing (until I find a better, less hacky, solution). It breaks when there is more than a simple child class like a mixin or something like that. I use a helper function to recursively search all base classes and find the generic one with all the types.

Also this syntax is technically not correct (I think; quite unsure to be honest): `update_data: update_schema`, because `update_schema` is a variable type while there should be an alias. It does seem to work though. Mypy doesn't like this but I switched to pyright because of the better support for generics.

Overall I'm quite happy with this even though it's not very clean. The ugly code is in a single place and the big advantage is that it greatly reduces code duplication.

Also, I don't see anything FastAPI could do to improve this.

**Reply by ahmed-aly-aut** [reply]
@tiangolo can you maybe take a look and comment your thoughts on this issue?

**Reply by bukacdan** [reply]
@tiangolo pinging this, having a built-in support for generics would be super useful
